### PR TITLE
plot: add sherpa_plot context manager

### PR DESCRIPTION
**Pull request by DougBurke**
# Summary

Add a sherpa_plot context manager to simplify creating a plot with the Sherpa plotting objects (only an issue if calling the plot objects directly, and not with commands like the UI plot_data call).

# Details

Creating a sherpa plot follows the same template, and it fits in well to being a context manager (even though there is no "plot object" we have to attach the open/close logic to, we can use the contextmanager decorator).

This is mainly useful internally, but it may be useful outside of Sherpa.

It should be separate from (but will conflict with) the plot-backend rework in #1382 / #1550. I guess we could add the backend-name as an optional argument to the context manager, to over-ride the backend for just that plot. Not sure how useful that would be, but it sounds cool!

**OOhhhhh** - a thought. We *could* make the backend provide a `sherpa_plot`-like routine, and then this version would just call that. Thoughts

- we could then remove the `setup`/`teardown`/`exceptions` logic as this would now just be handled within the plotend so if you don't need to do anything special with exception-handling you just don't do it
- normally you would add a context manager to a class and have it associated with the lifetime of that object, but that's not true here; is this confusing?

### CODE REVIEW

File: sherpa/plot/__init__.py [hunk 1]
```diff
@@ -49,8 +49,33 @@
   >>> import sherpa.plot.pylab_backend
   >>> plot.backend = sherpa.plot.pylab_backend
 
+Creating a plot
+---------------
+
+The simplest approach is to use the `sherpa_plot` context manager::
+
+    with sherpa_plot():
+        # Now call the plot/overplot or contor/overcontour methods
```
Comment (wmclaugh) [inline sherpa/plot/__init__.py:58]:
```suggestion
        # Now call the plot/overplot or contour/overcontour methods
```

File: sherpa/plot/__init__.py [hunk 2]
```diff
@@ -49,8 +49,33 @@
...
+        obj.plot()
+
+This handles setting up the backend, handles any error handling,
+and then ends the session. It is equivalent to::
+
+    from sherpa import plot
+    try:
+        plot.backend.begin()
+        # Now call the plot/overplot or contor/overcontour methods
```
Comment (wmclaugh) [inline sherpa/plot/__init__.py:67]:
```suggestion
        # Now call the plot/overplot or contour/overcontour methods
```

Comment (DougBurke) [reply sherpa/plot/__init__.py:67]: I'm thinking that this "equivalent to" section can just be dropped, as I think we want to make `sherpa_plot` (or whatever we decide to bike-shed the name to) would become the way to setup plots, but need to look at what other backends do (related to my excitement above in working out possible synergies with the backend rework).

I'll still take the change though!

**Comment by hamogu** [comment]
This is a slightly different angle to take than #1382 as it would be a change in the API of how to do a plot, which #1382 tries to avoid (not entirely, but as much as reasonable).

A context manager *is* the Pythoninc way to deal with the setup/teardown, so this is worth thinking about. I don't think is collides as much with #1382, as with later steps on where the plot preferences live. My view is that some preferences live in the plot object (symbols/lines for spectral data), while other might live in the backend (e.g. the default color map which will be named differently in matplotlib vs. bokeh, vs. altair). I've written up thoughts about that in multiple places and I'm not going to pull it together in this location again, but it's partially discussed in #1375. So, when I call this context manager, do I get a default version of the backend? Do I get a copy of the current settings? How do I persist settings if the backend is generated in the context manager? Or would I pass a backend *instance* (as opposed the the name/class) into the context manager, because the *instance* is were the user-set preferences live.

All that can be done, but I wonder if it's feature creep - in short, I suggest to keep this idea in the back of our minds at least until #1383 is in.

**Comment by hamogu** [comment]
I had forgotten about this PR and have in fact implemented something similar in https://github.com/hamogu/sherpa/commit/0be22d5ec690a1c63cdb6d5364ed8a0f6a327d83

**Comment by DougBurke** [comment]
@hamogu - do you want me to close this PR down then?

**Comment by DougBurke** [comment]
@hamogu - I plan to close this but should I keep it open?

**Pull request closed without merging**